examples/c/ex_file_system.c

```diff
@@ -0,0 +1,650 @@
+ *	fopen for our demo file system
+ */
+static int
+demo_fs_open(WT_FILE_SYSTEM *file_system, WT_SESSION *session,
+    const char *name, int file_type, u_int flags, WT_FILE_HANDLE **file_handlep)
+{
+	WT_FILE_HANDLE *file_handle;
+	DEMO_FILE_HANDLE *demo_fh;
+	DEMO_FILE_SYSTEM *demo_fs;
+
+	(void)file_type;					/* Unused */
+	(void)session;						/* Unused */
+	(void)flags;						/* Unused */
+
+	demo_fs = (DEMO_FILE_SYSTEM *)file_system;
+	demo_fh = NULL;
+
+	++demo_fs->opened_file_count;
+
+	/*
+	 * First search the file queue, if we find it, assert there's only a
+	 * single reference, we only supports a single handle on any file.
+	 */
+	demo_fh = demo_handle_search(file_system, name);
+	if (demo_fh != NULL) {
+		if (demo_fh->ref != 0) {
+			fprintf(stderr,
+			    "demo_file_open of already open file %s\n",
+			    name);
+			return (EBUSY);
+		}
+
+		demo_fh->ref = 1;
+		demo_fh->off = 0;
+
+		*file_handlep = (WT_FILE_HANDLE *)demo_fh;
+		return (0);
+	}
+
+	/* The file hasn't been opened before, create a new one. */
+	if ((demo_fh = calloc(1, sizeof(DEMO_FILE_HANDLE))) == NULL)
+		return (ENOMEM);
+
+	/* Initialize private information. */
+	demo_fh->ref = 1;
+	demo_fh->off = 0;
+	demo_fh->demo_fs = demo_fs;
+	if ((demo_fh->buf = calloc(1, DEMO_FILE_SIZE_INCREMENT)) == NULL)
+		goto enomem;
+	demo_fh->size = DEMO_FILE_SIZE_INCREMENT;
+
+	/* Initialize public information. */
+	file_handle = (WT_FILE_HANDLE *)demo_fh;
+	if ((file_handle->name = strdup(name)) == NULL)
+		goto enomem;
+
+	file_handle->close = demo_file_close;
+	file_handle->lock = demo_file_lock;
+	file_handle->read = demo_file_read;
+	file_handle->size = demo_file_size;
+	file_handle->truncate = demo_file_truncate;
+	file_handle->write = demo_file_write;
+
+	TAILQ_INSERT_HEAD(&demo_fs->fileq, demo_fh, q);
+	++demo_fs->opened_unique_file_count;
+
+	*file_handlep = file_handle;
+	return (0);
+
+enomem:	free(demo_fh->buf);
+	free(demo_fh);
+	return (ENOMEM);
+}
+
+/*
+ * demo_fs_directory_list --
+ *	Return a list of files in a given sub-directory.
+ */
+static int
+demo_fs_directory_list(WT_FILE_SYSTEM *file_system, WT_SESSION *session,
+    const char *directory, const char *prefix, char ***dirlistp, u_int *countp)
+{
+	DEMO_FILE_HANDLE *demo_fh;
+	DEMO_FILE_SYSTEM *demo_fs;
+	size_t dir_len, prefix_len;
+	char *name, **list;
+	u_int allocated, count;
+
+	(void)session;						/* Unused */
+
+	demo_fs = (DEMO_FILE_SYSTEM *)file_system;
+	list = NULL;
+	allocated = count = 0;
+	dir_len = strlen(directory);
+	prefix_len = prefix == NULL ? 0 : strlen(prefix);
+
+	TAILQ_FOREACH(demo_fh, &demo_fs->fileq, q) {
+		name = demo_fh->iface.name;
+		if (strncmp(name, directory, dir_len) == 0) {
+			if (prefix != NULL &&
+			    strncmp(name, prefix, prefix_len) != 0)
+				continue;
+
+			++count;
+			/*
+			 * Increase the list size in groups of 10, it doesn't
+			 * matter if the list is a bit longer than necessary.
+			 */
+			if (count > allocated) {
+				list = realloc(
```

Comment:
@keithbostic @michaelcahill Is it OK for this method to allocate memory that is later freed by WiredTiger? It is the only place in the API where we do that at the moment. If we allow this we will need to ensure that file system implementations use the WiredTiger memory allocation routines.

The alternative would be to add a dirlist_free method to the API.